### PR TITLE
let occ user:lastseen report latest logins

### PR DESCRIPTION
**Pull request by butonic**
## Description

This PR makes the uid param of `occ user:lastseen` obsolete and returns the n most recently logged in users limited by a `--limit` parameter (default is 10). It also now supports `--output`. For JSON output the users displayname and email er added to the object to make the output actually useful, eg:

``` json
$ php ./occ user:lastseen --output=json_pretty
[
    {
        "userid": "admin",
        "lastLogin": "2016-09-21T10:25:20",
        "displayname": "admin",
        "email": null
    },
    {
        "userid": "null",
        "lastLogin": "2016-09-05T09:29:58",
        "displayname": "null",
        "email": null
    }
]
```

Yes, I actually have a user `"null"` in this test db 😈 

The existing output when a uid is given remains unchanged.
## Related Issue

reporting functionality has been asked for quite often
## Motivation and Context

- [x] Who is actively using the instance? (most recent logins, the default)
- [x] Who no longer uses the instance? (least recent logins, TODO) 
## How Has This Been Tested?

On the cmdline. I am trying to get write a unit test ... but mocking a querybuilder is no fun ...
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by butonic** [comment]
@DeepDiver1975 @PVince81 how do I mock a QueryBuilder and remain sane? Or shoud I move the DB code to the Config classes? Add an interface? Or just use the db?

**Comment by PVince81** [comment]
Use the DB in this case. We never mock the query builder AFAIK.

**Comment by butonic** [comment]
- [ ] add option to sort by least recently logged in
- [ ] timestamps in db are UTC so add 'Z' in JSON format

**Comment by PVince81** [comment]
Should the output rather be in JSON or table format to make it easier to read/parse ?
Current output:
```
8993415c-1f22-1036-833a-2148c326cd92`s last login: 05.10.2016 08:38
8992da6e-1f22-1036-8339-2148c326cd92`s last login: 05.10.2016 08:37
admin`s last login: 05.10.2016 08:36
user7`s last login: 05.10.2016 08:35
user6`s last login: 05.10.2016 08:35
user4`s last login: 05.10.2016 08:34
user5`s last login: 05.10.2016 08:34
user3`s last login: 05.10.2016 08:31
User2`s last login: 05.10.2016 08:30
user1`s last login: 05.10.2016 08:30
```

**Comment by butonic** [comment]
```console
$ php occ user:lastseen --help
Usage:
  user:lastseen [options] [--] [<uid>]

Arguments:
  uid                    the username

Options:
      --output[=OUTPUT]  Output format (plain, json or json_pretty, default is plain) [default: "plain"]
      --least-recent     show least recently logged in users
  -h, --help             Display this help message
  -q, --quiet            Do not output any message
  -V, --version          Display this application version
      --ansi             Force ANSI output
      --no-ansi          Disable ANSI output
  -n, --no-interaction   Do not ask any interactive question
      --no-warnings      Skip global warnings, show command output only
  -10, --limit[=LIMIT]   limit to n users
  -v|vv|vvv, --verbose   Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug

Help:
 shows when users were logged in last time

$ php occ user:lastseen
admin`s last login: 10.10.2016 21:13
zombie300@example.org`s last login: 28.09.2016 13:42
zombie301`s last login: 28.09.2016 07:35
zombie300`s last login: 28.09.2016 07:28
user1`s last login: 21.09.2016 10:25
user2`s last login: 05.09.2016 09:29

$ php occ user:lastseen --least-recent
user2`s last login: 05.09.2016 09:29
user1`s last login: 21.09.2016 10:25
zombie300`s last login: 28.09.2016 07:28
zombie301`s last login: 28.09.2016 07:35
zombie300@example.org`s last login: 28.09.2016 13:42
admin`s last login: 10.10.2016 21:13

$ php occ user:lastseen --least-recent --output=json_pretty
[
    {
        "displayname": null,
        "email": null,
        "lastLogin": "2016-09-05T09:29:58Z",
        "userid": "user2"
    },
    {
        "displayname": null,
        "email": null,
        "lastLogin": "2016-09-21T10:25:20Z",
        "userid": "user1"
    },
    {
        "displayname": null,
        "email": null,
        "lastLogin": "2016-09-28T07:28:14Z",
        "userid": "zombie300"
    },
    {
        "displayname": null,
        "email": null,
        "lastLogin": "2016-09-28T07:35:14Z",
        "userid": "zombie301"
    },
    {
        "displayname": null,
        "email": null,
        "lastLogin": "2016-09-28T13:42:00Z",
        "userid": "zombie300@example.org"
    },
    {
        "displayname": "admin",
        "email": null,
        "lastLogin": "2016-10-10T21:13:38Z",
        "userid": "admin"
    }
]
```

**Comment by butonic** [comment]
- [ ] add option to list users that have not logged in in the last x days

**Comment by butonic** [comment]
@DeepDiver1975 @PVince81 @PhilippSchaffrath pls review

File: core/Command/User/LastSeen.php [hunk 1]
```diff
@@ -24,51 +24,199 @@
...
+							// user no longer exists
+							$users[$key]['displayname'] = null;
+							$users[$key]['email'] = null;
+						}
+						ksort($users[$key]);
+						break;
+					default:
+						$output->writeln(
+							$data['userid'] .'`s last login: ' . $date->format('d.m.Y H:i') );
+						break;
```
Comment (PhilippSchaffrath) [inline core/Command/User/LastSeen.php]: unnecessary

**Review by phisch — COMMENTED** [review]
Regarding the tests, can we agree on putting the expectation in the method name? It sometimes gets quite long that way, but figuring out why a test failed is much more easy that way.

**Comment by PVince81** [comment]
@butonic please address @PhilippSchaffrath's comments

**Comment by butonic** [comment]
cc @cdamken maybe you can test and tell me what you need
I already heard
- [ ] CSV output
- [ ] user backend (allows distinguishing guests from users)

File: core/Command/User/LastSeen.php [hunk 2]
```diff
@@ -24,51 +24,199 @@
...
+		$query = $queryBuilder->execute();
+		$result = [];
+
+		while ($row = $query->fetch()) {
+			if (!empty($row['configvalue']) && ($before === null || $row['configvalue'] <= $before )) {
+				$result[] = ['userid' => $row['userid'], 'lastLogin' => $row['configvalue']];
+			}
+		}
+
+		if (empty($limit) || count($result) >= $limit) {
```
Comment (butonic) [inline core/Command/User/LastSeen.php]: just defensive programming

**Comment by butonic** [comment]
should we just omit dasplayname and email if a user no longer exists and instead add a `"deleted":true`?

**Comment by butonic** [comment]
@jvillafanez @cdamken @dercorn can you comment on https://github.com/owncloud/core/pull/26184#issuecomment-272566074

**Comment by jvillafanez** [comment]
Taking into account LDAP, I'd keep the displayname and email showing. The new property should be available always, not just for deleted users.

Maybe instead of `"deleted": true` we can change it into `"status": "active"`, `"status": "missing"`... it can be extended easily.

**Comment by PVince81** [comment]
annnnd... when the user account table arrives will we also have last login there ?

**Comment by PVince81** [comment]
moving to triage for proper scheduling / time allocation

**Comment by tomneedham** [comment]
With the accounts table this is much easier as we have the `last_login` column. Needs to be reimplemented. Enjoyed the oracle magic though ;)

**Pull request closed without merging**